clang/lib/AST/ByteCode/InterpBuiltin.cpp

```diff
@@ -2817,6 +2817,82 @@ static bool interp__builtin_select(InterpState &S, CodePtr OpPC,
   return true;
 }
 
+static bool interp__builtin_elementwise_fsh(InterpState &S, CodePtr OpPC,
+                                            const CallExpr *Call,
+                                            unsigned BuiltinID) {
+  assert(Call->getNumArgs() == 3);
+
+  const QualType &Arg1Type = Call->getArg(0)->getType();
+  const QualType &Arg2Type = Call->getArg(1)->getType();
+  const QualType &Arg3Type = Call->getArg(2)->getType();
+
+  // Non-vector integer types.
+  if (!Arg1Type->isVectorType()) {
+    assert(!Arg2Type->isVectorType());
+    assert(!Arg3Type->isVectorType());
+    const APSInt &Shift = popToAPSInt(
+        S.Stk, *S.getContext().classify(Call->getArg(2)->getType()));
+    const APSInt &Lo = popToAPSInt(
+        S.Stk, *S.getContext().classify(Call->getArg(1)->getType()));
+    const APSInt &Hi = popToAPSInt(
+        S.Stk, *S.getContext().classify(Call->getArg(0)->getType()));
+    APSInt Result;
+    if (BuiltinID == Builtin::BI__builtin_elementwise_fshl) {
+      Result = APSInt(llvm::APIntOps::fshl(Hi, Lo, Shift), Hi.isUnsigned());
+    } else if (BuiltinID == Builtin::BI__builtin_elementwise_fshr) {
+      Result = APSInt(llvm::APIntOps::fshr(Hi, Lo, Shift), Hi.isUnsigned());
+    } else {
+      llvm_unreachable("Wrong builtin ID");
+    }
+    pushInteger(S, Result, Call->getType());
+    return true;
+  }
+
+  // Vector type.
+  assert(Arg1Type->isVectorType() && Arg2Type->isVectorType() &&
+         Arg3Type->isVectorType());
+
+  const VectorType *VecT = Arg1Type->castAs<VectorType>();
+  const PrimType &ElemT = *S.getContext().classify(VecT->getElementType());
+  unsigned NumElems = VecT->getNumElements();
+
+  assert(VecT->getElementType() ==
+             Arg2Type->castAs<VectorType>()->getElementType() &&
+         VecT->getElementType() ==
+             Arg3Type->castAs<VectorType>()->getElementType());
```

Comment:
1) Can you add these assertions in the current interpreter as well?
2) I _think_ they aren't exactly correct and will fail if one vector is of type int and the other of type `const int`. However, I'm not sure if that's even possible atm.

Reply:
These types are not considered equal much before constexpr evaluation:

```
typedef const short vector4cshort __attribute__((__vector_size__(8)));
typedef short vector4short __attribute__((__vector_size__(8)));
```


```
error: arguments are of different types ('vector4cshort' (vector of 4 'const short' values) vs 'vector4short' (vector of 4 'short' values))
   25 |   constexpr auto chk = __builtin_elementwise_fshl(x,y,z);
      |                                                     ^
```

Do you think I should still add these assertions to the current interpreter?

Reply:
Created #155405.